# Лабораторна робота 3

### Завдання 2

1. Прочитайте csv файл (метод read_csv)

In [1]:
import pandas as pd
import numpy as np

data = pd.read_csv("bestsellers with categories.csv", encoding='latin1')
data

Name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       Author  User Rating  Reviews  Price  Year        Genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[550 rows x 7 columns]

2. Виведіть перші 10 рядків (метод head) 

In [2]:
data.head(10)

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   
5      A Dance with Dragons (A Song of Ice and Fire)   
6  A Game of Thrones / A Clash of Kings / A Storm...   
7                     A Gentleman in Moscow: A Novel   
8      A Higher Loyalty: Truth, Lies, and Leadership   
9                          A Man Called Ove: A Novel   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction  
5       George R. R. Martin          4.4    12643     11  2011      Fiction  
6       George R. R. Martin          4.7    19735     30  2014      Fiction  
7               Amor Towles          4.7    19699     15  2017      Fiction  
8               James Comey          4.7     5983      3  2018  Non Fiction  
9           Fredrik Backman          4.6    23848      8  2016      Fiction

3. Виведіть розміри датасету (метод shape) 

In [3]:
data.shape

(550, 7)

Питання: Про скільки книг зберігає дані датасет? 

Для кожної з книг доступні 7 колонок.  
1) Name - назва книги 
2) Author - автор 
3) User Rating - рейтинг (за 5-бальною шкалою) 
4) Reviews - кількість відгуків 
5) Price - ціна (у доларах станом на 2020 рік) 
6) Year - рік, коли книга потрапила до рейтингу Топ-50 
7) Genre - жанр 

Відповідь: 550

4. Змініть назви всіх колонок: назва з маленької букви, без пробілів через підкреслення. 

df.columns = ['name', 'author', 'user_rating', 'reviews', 'price', 'year', 'genre'] 

In [4]:
data.columns = ['name', 'author', 'user_rating', 'reviews', 'price', 'year', 'genre'] 
data

name  \
0                        10-Day Green Smoothie Cleanse   
1                                    11/22/63: A Novel   
2              12 Rules for Life: An Antidote to Chaos   
3                               1984 (Signet Classics)   
4    5,000 Awesome Facts (About Everything!) (Natio...   
..                                                 ...   
545       Wrecking Ball (Diary of a Wimpy Kid Book 14)   
546  You Are a Badass: How to Stop Doubting Your Gr...   
547  You Are a Badass: How to Stop Doubting Your Gr...   
548  You Are a Badass: How to Stop Doubting Your Gr...   
549  You Are a Badass: How to Stop Doubting Your Gr...   

                       author  user_rating  reviews  price  year        genre  
0                    JJ Smith          4.7    17350      8  2016  Non Fiction  
1                Stephen King          4.6     2052     22  2011      Fiction  
2          Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3               George Orwell          4.7    21424      6  2017      Fiction  
4    National Geographic Kids          4.8     7665     12  2019  Non Fiction  
..                        ...          ...      ...    ...   ...          ...  
545               Jeff Kinney          4.9     9413      8  2019      Fiction  
546               Jen Sincero          4.7    14331      8  2016  Non Fiction  
547               Jen Sincero          4.7    14331      8  2017  Non Fiction  
548               Jen Sincero          4.7    14331      8  2018  Non Fiction  
549               Jen Sincero          4.7    14331      8  2019  Non Fiction  

[550 rows x 7 columns]

5. Вивести кількість пропусків у кожному стовпчику. 

In [5]:
data_miss = data.isna().sum()/len(data)*100
print("Кількість пропусків у кожному стовпчику: \n", data_miss)

Кількість пропусків у кожному стовпчику: 
 name           0.0
author         0.0
user_rating    0.0
reviews        0.0
price          0.0
year           0.0
genre          0.0
dtype: float64


Питання : Чи є в якихось змінних пропуски? (Так / ні) 

Відповідь: Ні

6. За допомогою методу unique  перевірте. Які є унікальні жанри в стовпці genre  

In [6]:
unique_genres = data['genre'].unique()
print("Унікальні жарни: ", unique_genres)

Унікальні жарни:  ['Non Fiction' 'Fiction']


Питання : Які є унікальні жанри? 

Відповідь: non fiction, fiction

7. Визначте максимальну, мінімальну, медіанну ціну за допомогою методів  max, min, mean, median 

In [7]:
max_price = data['price'].max()
min_price = data['price'].min()
mean_price = data['price'].mean()
median_price = data['price'].median()

print("Максимальна: ", max_price)
print("Мінімальна: ", min_price)
print("Середня: ", mean_price)
print("Медіанна: ", median_price)

Максимальна:  105
Мінімальна:  0
Середня:  13.1
Медіанна:  11.0


Питання: Максимальна ціна?

Відповідь: 105


Питання: Мінімальна ціна? 

Відповідь: 0


Питання: Середня ціна? 

Відповідь: 13,1


Питання: Медіанна ціна? 

Відповідь: 11

8. Виконайте наступні завдання з пошуку і сортування даних         

Питання: Який рейтинг у датасеті найвищий?  

In [8]:
max_rating = data['user_rating'].max()
print("Максимальний рейтинг: ", max_rating)

Максимальний рейтинг:  4.9


Питання: Скільки книг мають такий рейтинг?  

In [9]:
amount_books = data[data['user_rating'] == max_rating]['name'].count()
print("Кількість книг з рейтингом 4.9: ", amount_books)

Кількість книг з рейтингом 4.9:  52


Питання: Яка книга має найбільше відгуків?  

In [10]:
max_reviews = data['reviews'].idxmax()
name_of_book = data.loc[max_reviews, 'name']
print("Книга яка має найбільше відгуків: ", name_of_book)

Книга яка має найбільше відгуків:  Where the Crawdads Sing


Питання: З тих книг, що потрапили до Топ-50 у 2010 році, яка книга найдорожча ? 

In [11]:
tbook_max_2010 = data[(data['year'] == 2010) & (data['user_rating'] <= 50)]
expensive_book = tbook_max_2010.loc[tbook_max_2010['price'].idxmax()]
name = expensive_book['name']
price = expensive_book['price']

print("Найдорожча книга з топ-50 у 2010 році: ", name, " - ", price)

Найдорожча книга з топ-50 у 2010 році:  Publication Manual of the American Psychological Association, 6th Edition  -  46


Питання : Скільки книг жанру Fiction потрапили до Топ-50 у 2012 році  

In [13]:
book_2012 = data[(data['year'] == 2012) & (data['user_rating'] <= 50)]
fiction = len(book_2012[book_2012['genre'] == 'Fiction'])

print("Кількість книг жанру fiction, які потрапили до топ-50 у 2012 році: ", fiction)

Кількість книг жанру fiction, які потрапили до топ-50 у 2012 році:  21


Питання : Скільки книг з рейтингом 4.9 потрапило до рейтингу у 2010 та 2011 роках (використовуйте | або метод isin)?  

In [14]:
years = (data['year'] == 2010) | (data['year'] == 2011)
top_rating = data[years & (data['user_rating'] == 4.9)]
amount = len(top_rating)

print("Кількість книг з рейтингом 4.9, які потрапили до рейтингу 2010 та 2011 роках:  ", amount)

Кількість книг з рейтингом 4.9, які потрапили до рейтингу 2010 та 2011 роках:   1


9. Вивести максимальну і  мінімальну ціну для жанру Fiction I NonFiction (за допомогою методів  groupby та agg, для підрахунку мінімальних та максимальних значень використовуйте max та min). Використовуйте тількі потрібні вам стовпці. 

In [15]:
genre_group = data.groupby('genre')
result = genre_group['price'].agg(['min', 'max'])
print(result)

             min  max
genre                
Fiction        0   82
Non Fiction    0  105
